vg.1
.\"             -*-Nroff-*-
.\"
.TH VG 1 "29 Aug 2000" "" ""
.SH NAME
vg \- Edit an entry found by the cg script
.SH SYNOPSIS
.B vg
[ -l ] number
.SH DESCRIPTION
.B "vg "
is the companion script to cg(1) which fires off an editor to
edit an entry found by cg(1).  This script is run only with one
argument which just be an integer within rage of the output of
cg(1), as seen on the far left column.  (Such as "vg 0".)  Running
it with the
.B -l
option edits using the log from the last search.
.PP
.SH "COMMAND\-LINE OPTIONS"
.TP 0.5i
.B "\-l "
Edits from the last search rather than from this terminal
.PP
.SH FILES
.ne 3
.TP
.I ${HOME}/.cglast
Log file of the last search.
.ne 3
.TP
.I ${HOME}/.cgvgrc
Per-user configuration file (if the defaults are not desireable).
.fi

.SH "SEE ALSO"
.BR cg (1),
.BR perl (1),
.BR cscope (1)

.SH AUTHOR
.B vg
was written by Joshua Uziel <uzi@uzix.org>.


ChangeLog
----------------
Change Log
----------------
August 29, 2000
1.6.1 - Pretty much added two things... 

	* "Colon mode"... which prints out something like this:

41 cgvg-common.pl:94  open (IN, "<$RCFILE");

	  instead of this:

41 cgvg-common.pl  94 open (IN, "<$RCFILE");
	
	  written on the request of a coworker who wanted it to mesh with
	  his editor a bit better.

	* Thanks to the good grace of Jeffrey R. Hay, cgvg now how has
	  built-in pager so that colors will remain while paging and
	  editing can occur while viewing the output.  Check it out. :)

November 3, 1999
1.6.0 - Some decent change to this version.
	
	* Cleaned up the code some.  I've introduced a cgvg-common.pl
	  file.  This does (and will) contain common subroutines and
	  variables.  More stuff will move there as I possibly do a GUI.
	  (If I do one.)  The script are now also broken into subroutines
	  a bit for easier readability and managability.  Some more cleaning
	  has to happen, but this is a start.

	* Logfile storage has changed quite a bit.  Now instead of one
	  global .cglast file, I have a .cgvg/ directory where I store
	  a log file for each shell in the form $HOSTNAME.shell_pid.
	  This is so that a user can use cgvg in several different places
	  and not lose their searches.  The name is for cleanup.  If it
	  doesn't find a pid that there's a log for on the present machine,
	  it deletes the log file.  The last search is still obtainable, but
	  now with a '-l' option... so "cg -l" to see the last search, and
	  "vg -l number" to edit an entry from that log.  Also note that
	  not having a log with the present terminal will default it to
	  the last log.  (Some of this idea and the next two came from Goran
	  Koruga <goran.koruga@hermes.si>)
	
	* There is no more BOLD option.  Now, all colors are achieveable
	  as the color, and b_color for the bold version (so "red" and
	  "b_red" for example).
	
	* The EXCLUDE option is now in the logfile, as is an AGE option
	  for a default cleanup time on logfiles.

	* There is now a configure script for this program to make things
	  (hopefully) more convenient.  (Fix by Manish Singh <yosh@gimp.org>)
	
	* Fix for handling the $EDITOR variable better if the editor is
	  using any arugments.  (Fix by Clinton Popetz <cpopetz@cpopetz.com>)

	* Also searching for *.l, *.y and *.C files by default as well.
	  (Request by Adam Marks <adam@cp.net>)

October 1, 1999
1.5.2 - Just a small bugfix release.  First, I removed a debug statement
	that I forgot to take out before sending off the last version
	(oops).  The bigger change is that now it'll parse through
	directories that are given as arguments... so if you do a
	"cg 'pattern' dir1 dir2", it'll recursively search through
	those directories.

August 25, 1999
1.5.1 - Fixed a bug where if one and only one file was specified
	to search through and it was an existing text file that it
	would read that file rather than blindly assume it's a
	search expression and do a huge recursive search.

	Cleaned up the finding code a bit.  Now both the default
	search and a user-specified search use the same &wanted
	subroutine, with the search list being specified by a
	$SEARCH variable.  Along with this, the default search list
	is user-defineable in the .cgvgrc file as well.

August 18, 1999
1.5 -	The big change of this version is that now the searching is
	handled by Perl rather than by grep(1), egrep(1) and fgrep(1).
	This required quite a bit of change, but everything is working
	like it was before.  The only really desirable option beyond
	the '-n' option (which I've covered) was '-i', which is
	covered as the only option at the moment.  The only things
	that are not handled by Perl now and are hard-coded now are
	"stty" and "which", and when I find a clean way of implementing
	them without requiring a trip to CPAN, I'll add those as well.

	Bugs have been fixed with handling the configuration file,
	and with the user-specified search list.  Now a user can
	search (for example) all C files with "cg pattern *.c" or
	"cg pattern '*.c'"... with the former being expanded by the
	shell and the latter being pattern matched by Perl.  Also
	note that the latter is a recursive search whereas the former
	is only those files specified on the command-line.  Colors
	are also now configurable by the configuration file.

	Manpages and a Makefile have also been added for user convenience.

August 15, 1999
1.4 -	Updates made here fix a bug that not all of the search list
	was being utilized, and the biggest change being that cg
	now uses Perl's internal File::Find routine instead of an
	external /bin/find (or whatever) to get the list of files.

	Syntax has changed a bit because of this... when files to
	search are specified, it must be done in quotes (single or
	double) like "cg pattern '*'" to search all files.

	Also, there's now a ~/.cgvgrc file where users can specify
	their editor, and bold and coloring options so that not all
	system users have to use the default.  See the "cgvgrc_example"
	file for a sample with the defaults.  Currently only $EDITOR,
	$BOLD, $BOLD_ALTERNATE and $COLORS can be set.  A bug in
	version 1.3 was corrected with the $BOLD option as well.

July 7, 1999
1.3 -	Same as the last version, except now colors are supported
	in cg.  There are the options $BOLD, $BOLD_ALTERNATE and  
	$COLOR to be set to 1 (on) or 0 (off) if you do or don't 
	want these feature.  $COLOR is needed for the $BOLD
	options, and really you should select only one or neither
	of the two $BOLD options.  Colors only work when displaying
	straight to STDOUT.

	In vg, I've gotten rid of the $CGEDITOR variable and replaced
	it with the standard $EDITOR environment variable since I've
	found most editors to support the +number option.  Also fixed
	a bug to give a clear error when the editor doesn't exist.

June 30, 1999
1.2 -	More changes to the cg script.  Rather than simply print the
	logfile like previous versions, I actually attempt to make

vg.pl
# vg - Vi Grepped - edit a file at a line specified in a cg search.
# Copyright 1999 by Joshua Uziel <juziel@home.com> - version 1.5.2
#
# Usage: vg number
#
# Helper script to go with cg for opening a editor conveniently
# with the correct file and line as shown in cg's log.  Run this
# with a single numerical argument (ie. "vg 3") to edit the desired
# log entry.
#
# This script seems to work fine with a number of editors, including
# vi, vim, emacs, pico, joe, etc.  (Despite it's name.)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

# File where the log is.
$LOGFILE = "$ENV{'HOME'}/.cglast";

# Path to the rc file
$RCFILE = "$ENV{'HOME'}/.cgvgrc";

# Default editor to use.
$EDITOR = "vi";

# Use the $EDITOR environment variable if it exists.
$EDITOR = $ENV{'EDITOR'} if ($ENV{'EDITOR'});

# If the rc file exists, parse it and override the defaults.
if (-f $RCFILE) {
        open (IN, "<$RCFILE");

        while (<IN>) {
                chomp;
		
		# Strip spaces and skip blank and comment lines.
		s/^\s*//;
		next if (/^#/);
		next if (/^$/);

                ($key, $value) = split /=/;

                if ($key =~ /^EDITOR$/) {
                        $EDITOR=$value;
                } elsif ($key =~ /^BOLD$/) {
                        next;
                } elsif ($key =~ /^BOLD_ALTERNATE$/) {
                        next;
                } elsif ($key =~ /^SEARCH$/) {
                        next;
                } elsif ($key =~ /^COLOR[S1-4]$/) {
                        next;
                } else {
                        die "error: Unknown option '$key' in $RCFILE at line",
				"$..\n";
                }
        }

        close (IN);
}

# We can't edit if the editor doesn't exist.
`which $EDITOR`;
$no_editor = $?;
die "Error: Editor $EDITOR isn't in your path.\n" if ($no_editor);

# We need one argument only, and it has to be a number.
die "Error: Single argument needed.\n" unless ($#ARGV == 0);

# Set it to $num and test that it is a number.
$num = $ARGV[0];

die "Error: Non-numerical argument.\n" 
	unless (($num =~ /\d+/) && ($num !~ /\D+/));

die "Error: $LOGFILE doesn't exist.\n" unless (-f $LOGFILE);

open (IN, "<$LOGFILE");

# Check if our current path is the path when the log was taken.  If so,
# then nullify $path, else set it to $path/ so we can use anywhere.
$PWD="$ENV{PWD}\n";
$path=(split(/=/, <IN>))[1];
if ($PWD eq $path) {
	$path="";
} else {
	chomp($path);
	$path="$path/";
}

# Find the line we want and open up vi to that line.
foreach $line (<IN>) {
	($this, $file, $line) = (split(/:/, $line))[0,1,2];

	# Change this to execute the editor.
	exec "$EDITOR +$line $path$file" if ($num == $this);
}

# If we're here, there was a problem.
die "Error: Numerical argument invalid.\n";
